---
title: "Document for Variant Annotation"
author: "Yubin Xie"
date: "January 09, 2017"
output: 
  html_document:
    includes:
      before_body: header.html
    number_sections: yes
    theme: readable
    toc: yes
    toc_float: yes
---

**Last updated:** `r format(Sys.Date(),format="%B %d, %Y")`

#Introduction

Welocme to my document for variants annotation. In this documents, I listed worflow, tools and database inforamtion for variants annotation. So far, it supports:

+ Data acquisition and decryption
    + NCBI dbGaP data
    
+ VCF file processing
    + Multi-allelic line separation
    + Case and control samples separation
    
+ Annotation
    + Gene-, region- and filter-based annotation by ANNOVAR
    + Custom-made region-based annotion
    + Genename-based annotion
    + Allele counts/frequency calculation
    
+ Functional scripts
    + Annotation results binding

I will keep updating this document and I hope it can help you in your research.

#Data acquisition and decryption

**Background:** 

Genomic information contains individal health and identity information. Therefore measures such as authorized access, encryption are performed to ensure that individuals' privacy is protected. Here I take NCBI dbGaP data as an example.

**Method:** 

+ To begin with, supervisors, who are identified as principal investigators (PIs) by eRA Commons system, should request access by NCBI authorized access system (see detailed instruction at: [dbGaP Authorized Access](https://dbgap.ncbi.nlm.nih.gov/aa/wga.cgi?login=&page=login))

+ Once being granted access to the data you apply for, you can download it and then decrypt it. The desryption precess is complex. First of all, log in PI NCBI dbGaP account and download the specific dbGap repository key for the data that you requested. Then download the [SRA Toolkit](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=software). Next comes the crucial but diffcult step: SRA Toolkit configuration. Detailed offical instruction can be found at: [SRA Toolkit Documentation - Protected Data Usage Guide](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=dbgap_use). Here the only thing that I'd like to emphasize is that "your decryption commend is not valid if you current directory is not "import project directory", which is set during SRA Toolkit configuration"

#Multi-allelic line separation

**Background:** 

So far, you should be able to get access to the Variant Call Format (VCF) that contains structural variants, which most of the case, is compressed). A standard VCF file contains meta-information lines, a header line, and then data lines each containing information about a position in the genome. Genotype information of one or multiple samples for each position will be included in that specific line.

The old VCF formats required multiple alleles in the same line while latest VCF formats (4.1+) start to allow for a single loci to cover multiple rows when there are multiple alleles. Two standards are always the source of problems. During annotation, some tools, eg. annovar, will split multiple alleles in the one line into multiple rows automatically. While others, eg. vcftools, will keep multiple alleles in one line (at least, for latest version of these tools). So problems come when we try to combine annotation results that are generated by different tools. The best way to prevent such problem is to convert multiple alleles in one row into multiple rows.

**Method:** 

+ [BCFtools](http://www.htslib.org/download/) `norm` argument can achieve to convert multiple alleles in one row into multiple rows.

```
bcftools norm -m -both filename.vcf.gz -o outputfilename
```

`norm` function in BCFtools is designed to normalize indels in VCF files. It is able to Left-align and normalize indels, check if REF alleles match the reference, split multiallelic sites into multiple rows; recover multiallelics from multiple rows.` -m, --multiallelics -|+[snps|indels|both|any]` can split multiallelic sites into biallelic records (by add `-` after `-m`) or join biallelic sites into multiallelic records (`+`). The split or merged object can be specified by a string: `snps` or `indels` or `both` or`any`. For merging, if use `both` SNPs and indels will be merged separately into two records; if `any` SNPs and indels will be merged into a single record. As for spliting, I used `both` when spliting the multiple alleles. All the multiallelic sites are split into multiple rows. I have not tried any for spliting yet. See more details at [BCF Tools Offical documentation](http://www.htslib.org/doc/bcftools.html). Though another tool, VCFlib, can also do similar job, it will lost sample information of other alleles that in one loci. As a result, when couting total chromsomes of one specific allele, the chromsomes of sample with other alleles in the same loci will not be counted to total chromsomes. So I would recommand BCFtools.

#General annotation

##Gene- and filter-based annotation

>ANNOVAR is an efficient software tool to utilize update-to-date information to functionally annotate genetic variants detected from diverse genomes -- [ANNOVAR documentation](http://annovar.openbioinformatics.org/en/latest)

###Preparetion of input files
**Background:** 

ANNOVAR is a popular tool for variants annotation. It requires specific format, which is ANNOVAR input format (avinput). In that file, each line corresponds to one variant (no multiple alleles in one line. On each line, the first five space- or tab- delimited columns represent chromosome, start position, end position, the reference nucleotides and the observed nucleotides. Additonal columns are supported and will be included in the later process. Recently, VCF has been the most widely used "genotype calling" format. Transformation of format is needed.

**Method:** 

+ The `convert2annovar.pl` script in ANNOVAR can convert other "genotype calling" format, eg. VCF, into ANNOVAR input format. The `-format vcf4` can be used to convert VCF files into ANNOVAR input files.

```
perl convert2annovar.pl -format vcf4 infilename.vcf -outfile outfilename.avinput
```

In fact, the input file can also be `vcf.gz` format. The `convert2annovar.pl` script can calculate allelic frequency but not in a reasonal way. Allele frequency is defined as: (number of certain allele) / (total allele number). However, ANNOVAR calculates allelic frequency as: $(number of certain allele)/(total sample number *2)$. $Total sample number *2 = total chromsome number = total detected allele number + total missing genotype information allele number$. Sometimes the allele number that lacks genotype information takes a large proportion of total chromsome number, which makes ANNOVAR calculated frequency inaccurate. I will introduce better tools later that calculate allele frequency. In this case, we do not need sample information in our ANNOVAR annotation, which usually contains 1000+ columns and makes the file heavy. So you can crop the VCF file before annotation or do not keep sample information during format conversion. If you want to keep sample and other information besides the first 5 columns during conversion, add `-allsample -includeinfo`.

###Multi-annotation
Once the ANNOVAR input file is ready, we can start annotate the variants. ANNOVAR can perform gene-based annotation, region-based annotation and filter-based annotation with downloaded appropriate databases. Detailed annotation function and aviable database can be found in ANNOVAR documentation. First step is to download the database you need and then perform the annotation you want. 

**Method:** 

+ Here is an example from [ANNOVAR documentation](http://annovar.openbioinformatics.org/en/latest).

```
perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar refGene humandb
perl annotate_variation.pl -buildver hg19 -downdb cytoBand humandb/
perl annotate_variation.pl -buildver hg19 -downdb genomicSuperDups humandb/
perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar esp6500siv2_all humandb/
perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar 1000g2015aug humandb/
perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar exac03 humandb/
perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar avsnp147 humandb/ 
perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar dbnsfp30a humandb/
perl table_annovar.pl infilename.avinput humandb/ -buildver hg19 -out myanno -remove -protocol refGene,cytoBand,genomicSuperDups,esp6500siv2_all,1000g2015aug_all,1000g2015aug_eur,exac03,avsnp147,dbnsfp30a -operation g,r,r,f,f,f,f,f,f -nastring . -csvout
```

In practice, error came out when I performed the the last command line on my data. It did not specify the reason but I guess that it cound because I perform too mauch annotation on a large file (6746960 lines). So I deleted unnecessary annotation like all the region annotation and performed avsnp147 separately. You can always perform the annotation one by one and combine them later.

##Sample allele counts & frequency calculation

###Split samples into case and control group
**Background:** 

Sometimes, case and control samples are mixed in one VCF file, which requires preprocess to split them. Our Strategy is that 1) firstly get the name and order list of case and control sample respectively; 2) then split the VCF file into two files according to the information in previous list.

**Method:** 

+ I developed a [case-control.py](https://github.com/YubinXie/Variant_Annotation/blob/master/case-control.py) python script to achieve the whole process. In the python script, I firstly take case and control sample name list according their name pattern (in my file, case name starts with A while control B) from cropped VCF file which only contains header inforamtion. Then I call [BCFTools](http://www.htslib.org/doc/bcftools.html) `view` argument to split the target file in python.

```
bcftools view -s CaseNameString infilename.vcf.gz | bgzip -c > outfilename_case.vcf.gz
bcftools view -s ControlNameString infilename.vcf.gz | bgzip -c > outfilename_control.vcf.gz
```

Though VCF-subset tool can do similar thing, it is much slower than BCFTools. My vcf.gz file is 60 GB, vcf-subset can only process 600MB in 1 hour while BCFTools can handle around 10 GB in 1 hour.

###Calculate allele counts and frequency

**Method:** 

+ [VCFTools](https://vcftools.github.io/man_latest.html) can accurately calculate sample allele counts and frequency.

```
vcftools --gzvcf infilename.vcf.gz --freq —recode —recode-INFO-all --out outfilename_freq
vcftools --gzvcf infilename.vcf.gz --counts —recode —recode-INFO-all --out outfilename_counts
```

It took around 40 min for my PC to calculate counts/frequency for 60 GB vcf.gz file.

#Custom-made annotation

**Background:** 

Annotation with custom-made dataset sometimes is required for researchers. Here I list some methods based on my own experience.

##Region-based annotation by ANNOVAR

**Background:** Our lab member developed a exon constrain score matrix, which contains information of exon location and its constarin score. Region-based annotation of ANNOVAR can be used to annotate VCF file with that data. This function requires **bed** format as the input file, with at least first three cols being chrom, chromStart, chromEnd. 

**Method:** 

+ Code as following is used:

```
table_annovar.pl -buildver hg19 -protocol bed infilename.avinput -operation r -nastring . humandb/ -otherinfo -bedfile filename.bed -arg '-colsWanted 20’
```

In this code, the column information that you want to include in the annotation should be specified in `-arg '-colsWanted colnumber'` , where colnumber is the number of that column.


## Gene-name-based annotion
**Background:** 

Gene expression level data is from GTEx and is in GCT format.

**Method:** 

+ Annotation with [Gene expression in heart] by [GCTexpression_annotation.py]:

```
python GCTexpression_annotation.py GeneAnnotatedFile ExpressionDataFile OutputFile
```
Note: This script calculated the average gene expression of several organs listed in the file and annotated variants based on gene name. The GeneAnnotatedFile here is a CSV file that was annotated by ref.gene.  

#Integrating multi-annotation

**Background:** 

So far, we have generated multiple text-based annotated files. Our final step is to combine them into one file. 

**Method:** 

+ I developed a [BindingAnnotation.R](https://github.com/YubinXie/Variant_Annotation/blob/master/Annotation.R) R script that allows you to combine specific columns of chosen files into one file. You can simplely add or change the file and column number to achieve that.

#Supporting information

## Tools information
**[SRAToolkit]:** Tools to decrypt protected data from dbGaP.

**[VCFTools]:** Utilities for the variant call format (VCF) and binary variant call format (BCF).

**[BCFTools]:** BCFtools is a set of utilities that manipulate variant calls in the Variant Call Format (VCF) and its binary counterpart BCF. 

**[ANNOVAR]:** ANNOVAR is an efficient software tool to utilize update-to-date information to functionally annotate genetic variants detected from diverse genomes.

**[case-control.py]:** Python script to split samples into case and control group.

**[GCTexpression_annotation.py]:** Python script to do gene-name-based annotation, eg. gene expression.

**[BindingAnnotation.R]:** R script to bind differernt annotation results.

[SRAToolkit]:https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=std
[VCFTools]: https://vcftools.github.io/man_latest.html
[BCFTools]: http://www.htslib.org/doc/bcftools.html
[ANNOVAR]: http://annovar.openbioinformatics.org/en/latest/
[case-control.py]: https://github.com/YubinXie/Variant_Annotation/blob/master/case-control.py
[BindingAnnotation.R]: https://github.com/YubinXie/Variant_Annotation/blob/master/Annotation.R
[GCTexpression_annotation.py]: https://github.com/YubinXie/Variant_Annotation/blob/master/GCTexpression_annotation.py
## Annotation database
**[General annotation]:** From ANNOVAR. Database includes: refGene, esp6500siv2_all, 1000g2015aug, exac03, avsnp147, dbnsfp30a. Get the database by following code:
```
perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar refGene humandb
perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar esp6500siv2_all humandb/
perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar 1000g2015aug humandb/
perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar exac03 humandb/
perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar avsnp147 humandb/ 
perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar dbnsfp30a humandb
```
**[Missense Exon constrain score]:** From Nick. It is on the metric defined in page 29 of this [supplement](http://www.nature.com/nature/journal/v536/n7616/extref/nature19057-s1.pdf). Click the name and get the data from GitHub.

**[Gene expression in heart]:** From GTEx. The file contains the median RPKM by tissue. These medians were calculated from unnormalized RPKMs, directly from the file GTEx_Analysis_v6p_RNA-seq_RNA-SeQCv1.1.8_gene_rpkm.gct.gz. Click the name and download it from GTEx.

[General annotation]: http://annovar.openbioinformatics.org/en/latest/user-guide/startup/
[Gene expression in heart]: http://www.gtexportal.org/home/datasets
[Missense Exon constrain score]: https://github.com/YubinXie/AnnotateR/tree/master/Exac_Exon_Constraint


#Acknowledgement

Here I'd like to thank Nicholas Knoblauch for trouble-shooting on various probelms all the way.

# Contact
If you have any questions or suggestion, contact me by email: yoobintse@gmail.com. 
